firedrake/checkpointing.py

```diff
@@ -824,14 +952,23 @@ def save_function(self, f, idx=None, name=None):
                 _name = "_".join([PREFIX_EMBEDDED, f.name()])
                 _f = Function(_V, name=_name)
                 self._project_function_for_checkpointing(_f, f, method)
-                self.save_function(_f, idx=idx)
+                self.save_function(_f, idx=idx, t=t, timestep=timestep)
                 self.set_attr(path, PREFIX_EMBEDDED + "_function", _name)
             else:
                 # -- Save function topology --
                 path = self._path_to_function(tmesh.name, mesh.name, V_name, f.name())
                 self.require_group(path)
                 self.set_attr(path, PREFIX + "_vec", tf.name())
+                # store index or timestamp mode if in timestepping mode
                 self._save_function_topology(tf, idx=idx)
+                if idx is not None:
+                    self._set_timestep(
+                        mesh=mesh,
+                        name=f.name(),
+                        idx=idx,
+                        t=t,
+                        timestep=timestep
+                        )
```

Comment:
```suggestion
                if idx is not None:
                    t = CheckpointFile.DEFAULT_REAL if t is None else t
                    timestep = CheckpointFile.DEFAULT_REAL if timestep is None else timestep
                    try:
                        t = float(t)
                        timestep = float(timestep)
                    except ValueError:
                        raise ValueError(
                            """t and timestep must be convertible to float."""
                        )
                    has_time_attr = self.has_attr(path, PREFIX_TIMESTEPPING + "_times")
                    old_times = self.get_attr(path, PREFIX_TIMESTEPPING + "_times") if has_time_attr else []
                    old_timesteps = self.get_attr(path, PREFIX_TIMESTEPPING + "_timesteps") if has_time_attr else []
                    times = np.concatenate((old_times, [t]))
                    timesteps = np.concatenate((old_timesteps, [timestep]))
                    self.set_attr(path, PREFIX_TIMESTEPPING + "_times", times)
                    self.set_attr(path, PREFIX_TIMESTEPPING + "_timesteps", timesteps)
```
Moved from `_set_timestep`. Here we only save `_times` and `_timesteps`, and at the end of `save_function_topology`, we save `_indices`:
```
        ...
        if idx is not None:
            self.viewer.popTimestepping()
            has_time_attr = self.has_attr(path, PREFIX_TIMESTEPPING + "_indices")
            old_indices = self.get_attr(path, PREFIX_TIMESTEPPING + "_indices") if has_time_attr else []
            indices = np.concatenate((old_indices, [idx]))
            self.set_attr(path, PREFIX_TIMESTEPPING + "_indices", indices)
```
as I believe that `_indices` should be a topological property.